### DISCUSSION TITLE
Bedrock Novo with_structured_output error

### DISCUSSION THREAD

**Discussion by lalit2001** [langchain-ai/langchain-aws#312] [Q&A]
Hello,

I'm using ChatBedrock with this code

ChatBedrock(model = "amazon.nova-micro-v1:0", aws_access_key_id = creds_conf.AWS_ACCESS_KEY_ID,
                       aws_secret_access_key = creds_conf.AWS_SECRET_ACCESS_KEY,
                       aws_session_token = creds_conf.AWS_SESSION_TOKEN, region_name = creds_conf.REGION)

chain = system_prompt_template | self.llm.with_structured_output(SummaryResponseDTO)
res = chain.invoke({"data": str(data), "metric": metric, "formula": formula, "date":date_range})

and getting the below error 

botocore.exceptions.ParamValidationError: Parameter validation failed:
Invalid length for parameter toolConfig.tools[0].toolSpec.description, value: 0, valid min length: 1

**Comment by 3coins** [comment]
@lalit2001 
I believe `ChatBedrockConverse` should work correctly in this case. Here is a related issue #435, where I have added sample code that works.

**Comment by code-by-mahereddy** [comment]
This error occurs because Amazon Nova models do not support forced tool calling, which `with_structured_output` relies on by default.

The `ParamValidationError` from botocore means the Converse API is receiving parameters that Nova doesn't support.

## Solution: Use json_mode instead

```python
from langchain_aws import ChatBedrock
from pydantic import BaseModel

class SummaryResponseDTO(BaseModel):
    summary: str
    key_points: list[str]

llm = ChatBedrock(
    model="amazon.nova-micro-v1:0",
    region_name=creds_conf.REGION
)

# Use json_mode instead of tool_calling
chain = system_prompt_template | llm.with_structured_output(
    SummaryResponseDTO,
    method="json_mode"
)
result = chain.invoke({"data": str(data), "metric": metric, "formula": formula, "date": date_range})
```

## Why this happens

By default, `with_structured_output` uses `method="tool_calling"` which forces the model to call a tool with a schema matching your Pydantic model. Amazon Nova models don't support forced tool calling via the Bedrock Converse API.

Using `method="json_mode"` instead tells the model to output JSON directly (without tool calling), which Nova supports.

## Alternative: Use a parsing approach

If `json_mode` doesn't give you the exact schema you need, you can also use the PydanticOutputParser:

```python
from langchain_core.output_parsers import PydanticOutputParser

parser = PydanticOutputParser(pydantic_object=SummaryResponseDTO)
chain = system_prompt_template | llm | parser
```

**Comment by kimnamu** [comment]
Thanks @lalit2001 for pasting the exact botocore message, which made this easy to pin down, and @3coins and @code-by-mahereddy for the earlier answers.

The empty `toolSpec.description` comes from `SummaryResponseDTO` having no docstring: before langchain-aws 0.2.13 the description was passed through as is, and for a model without a docstring it is an empty string, which botocore rejects before sending (the Converse API requires at least one character). #347 (for #336, first released in 0.2.13) falls back to the schema name, so the original `ChatBedrock` code works unchanged on 0.2.13 and later; on an older version, a one-line docstring on the model is enough.

| Code from this thread | 0.2.10 (current in Dec 2024) | 0.2.13 | main `5649a8d`, live Nova Micro, us-east-1 |
|---|---|---|---|
| `ChatBedrock(model="amazon.nova-micro-v1:0").with_structured_output(SummaryResponseDTO)` | ❌ same `ParamValidationError` | ✅ | ✅ returns `SummaryResponseDTO` |
| same with `ChatBedrockConverse` | ❌ same error | ✅ | ✅ |
| docstring added to the model | ✅ | ✅ | ✅ |
| `with_structured_output(..., method="json_mode")` | ❌ same error | ✅ | ✅, still a forced `toolChoice` |

To add to @3coins's answer: what fixes it is the upgrade rather than the class, since `ChatBedrock` already sends Nova models through `ChatBedrockConverse`, and on 0.2.10 both fail the same way. On `method="json_mode"`: 0.2.10 ignores `method` here, and on main any value other than `json_schema` or `prompt_prefill` falls through to tool calling, so the request does not change. Forced tool calling was not the cause either: 0.2.10 set no `toolChoice` for Nova, and on main Nova Micro accepted `{"tool": {"name": ...}}` in us-east-1 (the only region I tried).

```
langchain-aws 0.2.10 / langchain-core 0.3.28 / botocore 1.35.86 / offline / amazon.nova-micro-v1:0 us-east-1
- ChatBedrock, no docstring (question)
    result: ParamValidationError: Parameter validation failed: Invalid length for parameter toolConfig.tools[0].toolSpec.description, value: 0, valid min length: 1
    request: {'description': [''], 'toolChoice': None}

langchain-aws 1.7.9 / langchain-core 1.6.4 / botocore 1.43.99 / live / amazon.nova-micro-v1:0 us-east-1
- ChatBedrock, no docstring (question)
    result: OK -> SummaryResponseDTO({"summary": "Q3 revenue rose 12% while costs stayed flat.", "key_point...)
    request: {'description': ['SummaryResponseDTO'], 'toolChoice': {'tool': {'name': 'SummaryResponseDTO'}}}
```

<details><summary>How this was run, full output, source pointers</summary>

0.2.10 and 0.2.13 were installed with dependencies as of their release dates (`uv pip install --exclude-newer 2024-12-23 langchain-aws==0.2.10`, `--exclude-newer 2025-02-19 langchain-aws==0.2.13`) and run offline with dummy credentials. The error is botocore's client-side parameter validation, so it is raised before anything is sent; the script also replaces the HTTP send, so nothing left the machine. main (reports 1.7.9) was run against Bedrock. `request` is what langchain-aws passed to `Converse`, recorded by wrapping `botocore.client.BaseClient._make_api_call`.

```python
class SummaryResponseDTO(BaseModel):  # no docstring, so the tool description is ""
    summary: str
    key_points: list[str]


class DocumentedDTO(BaseModel):
    """Summary of the metric data."""

    summary: str
    key_points: list[str]


MODEL, REGION = "amazon.nova-micro-v1:0", "us-east-1"
PROMPT = "Summarize: Q3 revenue rose 12% while costs stayed flat."
CASES = [
    ("ChatBedrock, no docstring (question)", lambda: ChatBedrock(model=MODEL, region_name=REGION).with_structured_output(SummaryResponseDTO)),
    ("ChatBedrockConverse, no docstring", lambda: ChatBedrockConverse(model=MODEL, region_name=REGION).with_structured_output(SummaryResponseDTO)),
    ("ChatBedrock, with docstring", lambda: ChatBedrock(model=MODEL, region_name=REGION).with_structured_output(DocumentedDTO)),
    ('ChatBedrock, method="json_mode"', lambda: ChatBedrock(model=MODEL, region_name=REGION).with_structured_output(SummaryResponseDTO, method="json_mode")),
]
# each case: build().invoke(PROMPT)
```

```
langchain-aws 0.2.10 / langchain-core 0.3.28 / botocore 1.35.86 / offline / amazon.nova-micro-v1:0 us-east-1
- ChatBedrock, no docstring (question)
    result: ParamValidationError: Parameter validation failed: Invalid length for parameter toolConfig.tools[0].toolSpec.description, value: 0, valid min length: 1
    request: {'description': [''], 'toolChoice': None}
- ChatBedrockConverse, no docstring
    result: ParamValidationError: Parameter validation failed: Invalid length for parameter toolConfig.tools[0].toolSpec.description, value: 0, valid min length: 1
    request: {'description': [''], 'toolChoice': None}
- ChatBedrock, with docstring
    result: OK -> passed client-side validation; HTTP send skipped (offline)
    request: {'description': ['Summary of the metric data.'], 'toolChoice': None}
- ChatBedrock, method="json_mode"
    result: ParamValidationError: Parameter validation failed: Invalid length for parameter toolConfig.tools[0].toolSpec.description, value: 0, valid min length: 1
    request: {'description': [''], 'toolChoice': None}

langchain-aws 0.2.13 / langchain-core 0.3.36 / botocore 1.36.23 / offline / amazon.nova-micro-v1:0 us-east-1
- ChatBedrock, no docstring (question)
    result: OK -> passed client-side validation; HTTP send skipped (offline)
    request: {'description': ['SummaryResponseDTO'], 'toolChoice': None}
- ChatBedrockConverse, no docstring
    result: OK -> passed client-side validation; HTTP send skipped (offline)
    request: {'description': ['SummaryResponseDTO'], 'toolChoice': None}
- ChatBedrock, with docstring
    result: OK -> passed client-side validation; HTTP send skipped (offline)
    request: {'description': ['Summary of the metric data.'], 'toolChoice': None}
- ChatBedrock, method="json_mode"
    result: OK -> passed client-side validation; HTTP send skipped (offline)
    request: {'description': ['SummaryResponseDTO'], 'toolChoice': None}

langchain-aws 1.7.9 / langchain-core 1.6.4 / botocore 1.43.99 / live / amazon.nova-micro-v1:0 us-east-1
- ChatBedrock, no docstring (question)
    result: OK -> SummaryResponseDTO({"summary": "Q3 revenue rose 12% while costs stayed flat.", "key_point...)
    request: {'description': ['SummaryResponseDTO'], 'toolChoice': {'tool': {'name': 'SummaryResponseDTO'}}}
- ChatBedrockConverse, no docstring
    result: OK -> SummaryResponseDTO({"summary": "Q3 revenue rose 12% while costs stayed flat", "key_points...)
    request: {'description': ['SummaryResponseDTO'], 'toolChoice': {'tool': {'name': 'SummaryResponseDTO'}}}
- ChatBedrock, with docstring
    result: OK -> DocumentedDTO({"summary": "Q3 revenue rose 12% while costs stayed flat.", "key_point...)
    request: {'description': ['Summary of the metric data.'], 'toolChoice': {'tool': {'name': 'DocumentedDTO'}}}
- ChatBedrock, method="json_mode"
    result: OK -> SummaryResponseDTO({"summary": "Q3 revenue increased by 12% while costs remained unchange...)
    request: {'description': ['SummaryResponseDTO'], 'toolChoice': {'tool': {'name': 'SummaryResponseDTO'}}}
```

Source, main `5649a8d`: the fallback is `tool_spec["description"] = tool_spec.get("description") or tool_spec["name"]` in `_format_tools` (`libs/aws/langchain_aws/chat_models/bedrock_converse.py` L3532), added by #347 on 2025-02-12; it is in the 0.2.13 wheel and not in 0.2.12. `ChatBedrock` turns on the Converse path for `amazon.nova` IDs in `set_beta_use_converse_api` (`chat_models/bedrock.py` L914-925; in 0.2.10 it was `"nova" in model_id`). The `method` dispatch is `bedrock_converse.py` L1760-1770, and 0.2.10's `with_structured_output` took `**kwargs` without reading them.
</details>
